Aircraft Generator

german-e-mas opened the pull request:
In this PR I include the random Aircraft generator.
The generation task is fired between 1 and 2 seconds, creating a random type of Aircraft in a random position near the edges of the Map (it wouldn't make sense if they appeared inside, out of nowhere). Their direction is set according to this as well.

I also made some changes in the Map class. It had a lot of static methods. I now instantiate the class and set it's boundaries in the constructor. The method to check if an Aircraft is inside the map is public, but the one to check for `Position` is private. I think it's more important to ask if a certain object is inside the map and have it's position be handled internally. Or having a single public method to check for position, and pass the object's position to it?

ptal @pablisho

pablisho commented on app/src/main/java/mas/german/landingplanes/AircraftGenerator.java:
```diff
@@ -0,0 +1,133 @@
+        mOnAircraftGeneratedListener = listener;
+    }
+
+    AircraftGenerator(Map currentMap) {
```
I would pass the Game here. At least for me it looks better, then inside the method you can ask for the map and store it

pablisho commented on app/src/main/java/mas/german/landingplanes/AircraftGenerator.java:
```diff
@@ -0,0 +1,133 @@
+        double x = 0;
+        double y = 0;
+        double angle = 0;
+        switch (3) {
```
switch 3?

german-e-mas replied:
I missed this correction. I was testing with fixed numbers, to see different creations in one particular side. It should be `mRandom.nextInt(4)`

pablisho commented on app/src/main/java/mas/german/landingplanes/AircraftGenerator.java:
```diff
@@ -0,0 +1,133 @@
+            default:
+                randomAircraft = new LargePlane(0, angle, randomPosition);
+        }
+        return randomAircraft;
```
What about velocity? All the planes have the same?

german-e-mas replied:
The speed is the first parameter on their creator. It's a random number between a minimum and a maximum, depending on the type. Should these min/max values be a static parameter of each class?

pablisho commented on app/src/main/java/mas/german/landingplanes/AircraftGenerator.java:
```diff
@@ -0,0 +1,133 @@
+    /**
+     * Schedules the first task. The task re-schedules itself after some time defined at random.
+     */
+    public void begin() {
```
it should have a stop too

pablisho commented on app/src/main/java/mas/german/landingplanes/Game.java:
```diff
@@ -108,4 +115,12 @@ private void gameOver() {
+    @Override
+    public void onAircraftGenerated(Aircraft generatedAircraft) {
+        // Synchronize the Aircraft list to prevent access during the operation.
+        synchronized (mAircrafts) {
```
This is ok, but you should also synchronize when accessing this list for reading. Just having one synchronize doesn't prevent the access, unless the other one also has a synchronize block on the same field

pablisho commented on app/src/main/java/mas/german/landingplanes/Map.java:
```diff
@@ -12,36 +12,57 @@
      * @return Map instance.
      */
-    public static Map getInstance() {
+    public static Map getInstance(double left, double top, double right, double bottom) {
```
This should definitely stop being a singleton

german-e-mas commented:
I addressed the comments.
- The list `mAircrafts` is synchronized every time something needs access to it.
- The default random Aircraft is `null`. The following switch statement will modify it.
- Map class is no longer singleton.
- Renamed the static final radius of each aircraft. There was no need to state the type in it's name.
- Included minimum and maximum speed values for each type of aircraft as static final integers.
- Included a `stop()` method for the Aircraft generator class, which shuts it down. The `begin()` method can restart the executor thread if it's called again after a shut down.
- Game's `gameOver()` method stops the generation of aircrafts with the previously mentioned method.

ptal @pablisho

pablisho commented on app/src/main/java/mas/german/landingplanes/aircrafts/Helicopter.java:
```diff
@@ -8,11 +8,14 @@
  */
 public class Helicopter extends Aircraft {
+    public static final int MAX_SPEED = 5;
+    public static final int MIN_SPEED = 5;
```
Always the same?

pablisho commented on app/src/main/java/mas/german/landingplanes/aircrafts/LargePlane.java:
```diff
@@ -8,11 +8,14 @@
  */
 public class LargePlane extends Aircraft {
+    public static final int MAX_SPEED = 10;
+    public static final int MIN_SPEED = 10;
```
same

pablisho commented on app/src/main/java/mas/german/landingplanes/Map.java:
```diff
@@ -3,45 +3,52 @@
-            sInstance = new Map();
-        }
-        return sInstance;
+    Map(double left, double top, double right, double bottom) {
```
Are you sure this names are ok? I mean, do you remember how we decided the coordinate system to be? Y up or down? if Y is down then up boundary should be 0 and down boundary should be 100

german-e-mas commented:
I addressed the comments about the speed and commented where the Map's origin is. Thanks for the review @pablisho , I'll merge now.
